src/app/ReadHandler.h

```diff
@@ -0,0 +1,134 @@
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *     This file defines read handler for a CHIP Interaction Data model
+ *
+ */
+
+#pragma once
+
+#include <app/InteractionModelDelegate.h>
+#include <core/CHIPCore.h>
+#include <core/CHIPTLVDebug.hpp>
+#include <messaging/ExchangeContext.h>
+#include <messaging/ExchangeMgr.h>
+#include <messaging/Flags.h>
+#include <protocols/Protocols.h>
+#include <support/CodeUtils.h>
+#include <support/DLLUtil.h>
+#include <support/logging/CHIPLogging.h>
+#include <system/SystemPacketBuffer.h>
+
+namespace chip {
+namespace app {
+/**
+ *  @class ReadHandler
+ *
+ *  @brief The read handler is responsible for processing read request, asks reporting engine to pack report data and sending
+ * reports via this handler.
+ *
+ */
+class ReadHandler
+{
+public:
+    /**
+     *  Initialize the ReadHandler. Within the lifetime
+     *  of this instance, this method is invoked once after object
+     *  construction until a call to Shutdown is made to terminate the
+     *  instance.
+     *
+     *  @param[in]    apExchangeMgr    A pointer to the ExchangeManager object.
+     *  @param[in]    apDelegate       InteractionModelDelegate set by application.
+     *
+     *  @retval #CHIP_ERROR_INCORRECT_STATE If the state is not equal to
+     *          kState_NotInitialized.
+     *  @retval #CHIP_NO_ERROR On success.
+     *
+     */
+    CHIP_ERROR Init(Messaging::ExchangeManager * apExchangeMgr, InteractionModelDelegate * apDelegate);
+
+    /**
+     *  Shutdown the ReadHandler. This terminates this instance
+     *  of the object and releases all held resources.
+     *  Shutdown is called internally when read handler fails to process read request, and fail or succeed to send report data
+     */
+    void Shutdown();
+    /**
+     *  Process read request
+     *
+     *  @param[in]    apExchangeContext    A pointer to the ExchangeContext.
+     *  @param[in]    aPayload             A payload that has read request data
+     *
+     *  @retval #Others If fails to process read request
+     *  @retval #CHIP_NO_ERROR On success.
+     *
+     */
+    CHIP_ERROR OnReadRequest(Messaging::ExchangeContext * apExchangeContext, System::PacketBufferHandle aPayload);
+
+    /**
+     *  Process read request
+     *
+     *  @param[in]    apExchangeContext    A pointer to the ExchangeContext.
+     *  @param[in]    aPayload             A payload that has read request data
+     *
+     *  @retval #Others If fails to process read request
+     *  @retval #CHIP_NO_ERROR On success.
+     *
+     */
+    CHIP_ERROR SendReportData(System::PacketBufferHandle aPayload);
+
+    bool IsFree() const { return mState == HandlerState::Uninitialized; }
+    bool IsReportable(void) { return mState == HandlerState::Initialized; }
+    bool IsReporting(void) { return mState == HandlerState::ReportDataSending; }
```

Comment:
Where are IsReportable and IsReporting used?

Reply:
Just removed, it would be used in ReportingEngine Run, I just remove these two PR, we could visit in IM eventing PR